File: Sprint-1/JavaScript/hasPairWithSum/hasPairWithSum.js
/**
 * Find if there is a pair of numbers that sum to a given target value.
 *
 * Time Complexity:
 * Space Complexity:
 * Optimal Time Complexity:
 *
 * @param {Array<number>} numbers - Array of numbers to search through
 * @param {number} target - Target sum to find
 * @returns {boolean} True if pair exists, false otherwise
 */
export function hasPairWithSum(numbers, target) {
  const seen = new Set()

  for (const number of numbers) {
      const complement = target - number

      if (seen.has(complement)) {
        return true;
      }
      seen.add(number)
  }
  return false;
}

File: Sprint-1/Python/has_pair_with_sum/has_pair_with_sum.py
from typing import List, TypeVar

Number = TypeVar("Number", int, float)


def has_pair_with_sum(numbers: List[Number], target_sum: Number) -> bool:
    """
    Find if there is a pair of numbers that sum to a target value.

    Time Complexity:
    Space Complexity:
    Optimal time complexity:
    """
    seen = set()

    for i in numbers:
        complement = target_sum - i
        if complement in seen:
            return True
        seen.add(i)
    return False
